File: sdk/python/libravdb/filters.py
from typing import List, Any, Optional

class Filter:
    def to_json(self) -> dict:
        raise NotImplementedError()

class Eq(Filter):
    def __init__(self, field: str, value: Any):
        self.field = field
        self.value = value
        
    def to_json(self) -> dict:
        return {"type": "eq", "field": self.field, "value": self.value}

class Gt(Filter):
    def __init__(self, field: str, value: Any):
        self.field = field
        self.value = value
        
    def to_json(self) -> dict:
        return {"type": "gt", "field": self.field, "value": self.value}

class Lt(Filter):
    def __init__(self, field: str, value: Any):
        self.field = field
        self.value = value
        
    def to_json(self) -> dict:
        return {"type": "lt", "field": self.field, "value": self.value}

class Between(Filter):
    def __init__(self, field: str, min_val: Any, max_val: Any):
        self.field = field
        self.min_val = min_val
        self.max_val = max_val
        
    def to_json(self) -> dict:
        return {"type": "between", "field": self.field, "value": [self.min_val, self.max_val]}

class ContainsAny(Filter):
    def __init__(self, field: str, values: List[Any]):
        self.field = field
        self.values = values
        
    def to_json(self) -> dict:
        return {"type": "contains_any", "field": self.field, "values": self.values}

class ContainsAll(Filter):
    def __init__(self, field: str, values: List[Any]):
        self.field = field
        self.values = values
        
    def to_json(self) -> dict:
        return {"type": "contains_all", "field": self.field, "values": self.values}

class ExactMatch(Filter):
    def __init__(self, field: str, values: List[Any]):
        self.field = field
        self.values = values
        
    def to_json(self) -> dict:
        return {"type": "exact_match", "field": self.field, "values": self.values}

class And(Filter):
    def __init__(self, *filters: Filter):
        self.filters = filters
        
    def to_json(self) -> dict:
        return {"type": "and", "filters": [f.to_json() for f in self.filters]}

class Or(Filter):
    def __init__(self, *filters: Filter):
        self.filters = filters
        
    def to_json(self) -> dict:
        return {"type": "or", "filters": [f.to_json() for f in self.filters]}

class Not(Filter):
    def __init__(self, filter: Filter):
        self.filter = filter
        
    def to_json(self) -> dict:
        return {"type": "not", "filter": self.filter.to_json()}

File: sdk/cgo/filters.go
package main

import (
	"encoding/json"
	"fmt"

	"github.com/xDarkicex/libravdb/internal/filter"
)

type ASTNode struct {
	Type    string        `json:"type"`
	Field   string        `json:"field,omitempty"`
	Value   interface{}   `json:"value,omitempty"`
	Values  []interface{} `json:"values,omitempty"`
	Filters []ASTNode     `json:"filters,omitempty"`
	Filter  *ASTNode      `json:"filter,omitempty"` // for NOT filter
}

func parseFilterJSON(jsonStr string) (filter.Filter, error) {
	if jsonStr == "" || jsonStr == "{}" {
		return nil, nil
	}

	var root ASTNode
	if err := json.Unmarshal([]byte(jsonStr), &root); err != nil {
		return nil, fmt.Errorf("failed to parse filter json: %v", err)
	}

	return buildFilter(root)
}

func buildFilter(node ASTNode) (filter.Filter, error) {
	switch node.Type {
	case "eq":
		return filter.NewEqualityFilter(node.Field, node.Value), nil
	case "gt":
		return filter.NewGreaterThanFilter(node.Field, node.Value), nil
	case "lt":
		return filter.NewLessThanFilter(node.Field, node.Value), nil
	case "between":
		// Assume Value is [min, max]
		arr, ok := node.Value.([]interface{})
		if !ok || len(arr) != 2 {
			return nil, fmt.Errorf("between filter requires an array of 2 values")
		}
		return filter.NewBetweenFilter(node.Field, arr[0], arr[1]), nil
	case "contains_any":
		return filter.NewContainsAnyFilter(node.Field, node.Values), nil
	case "contains_all":
		return filter.NewContainsAllFilter(node.Field, node.Values), nil
	case "exact_match":
		return filter.NewExactMatchFilter(node.Field, node.Values), nil
	case "and":
		var subFilters []filter.Filter
		for _, sub := range node.Filters {
			f, err := buildFilter(sub)
			if err != nil {
				return nil, err
			}
			subFilters = append(subFilters, f)
		}
		return filter.NewAndFilter(subFilters...), nil
	case "or":
		var subFilters []filter.Filter
		for _, sub := range node.Filters {
			f, err := buildFilter(sub)
			if err != nil {
				return nil, err
			}
			subFilters = append(subFilters, f)
		}
		return filter.NewOrFilter(subFilters...), nil
	case "not":
		if node.Filter == nil {
			return nil, fmt.Errorf("not filter missing child filter")
		}
		f, err := buildFilter(*node.Filter)
		if err != nil {
			return nil, err
		}
		return filter.NewNotFilter(f), nil
	default:
		return nil, fmt.Errorf("unknown filter type: %s", node.Type)
	}
}
